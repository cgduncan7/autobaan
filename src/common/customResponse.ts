import {
	Injectable,
	NestInterceptor,
	ExecutionContext,
	CallHandler,
} from '@nestjs/common'
import { Observable } from 'rxjs'
import { map } from 'rxjs/operators'

export interface CustomResponse<T = unknown> {
	data: T
}

@Injectable()
export class CustomResponseTransformInterceptor<T>
	implements NestInterceptor<T, CustomResponse<T>>
{
	intercept(
		_context: ExecutionContext,
		next: CallHandler,
	): Observable<CustomResponse<T>> {
		return next.handle().pipe(map((data) => ({ data })))
	}
}
