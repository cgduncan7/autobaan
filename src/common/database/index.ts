import mysql, { Connection, ConnectionConfig, FieldInfo } from 'mysql'
import { readFile } from 'fs/promises'
import { resolve } from 'path'
import { TABLE_reservations } from './sql'

const createConnectionConfig = async (): Promise<ConnectionConfig> => {
  const user = await readFile(resolve('.', './secrets/dbUser'))
  const password = await readFile(resolve('.', './secrets/dbPassword'))
  return {
    user: user.toString(),
    password: password.toString(),
    database: 'autobaan',
  }
}

let connection: Connection

export const getConnection = async (): Promise<Connection> => {
  if (!connection) {
    const config = await createConnectionConfig()
    connection = mysql.createConnection(config)
  }
  return connection
}

export type ResultSet<T> = T[]

export const connect = async () => {
  return new Promise<void>((res, rej) =>
    getConnection().then((cn) => {
      cn.connect((err) => {
        if (err) {
          rej(err)
        }
        res()
      })
    })
  )
}

export const disconnect = async () => {
  return new Promise<void>((res, rej) =>
    getConnection().then((cn) => {
      cn.end((err) => {
        if (err) {
          rej(err)
        }
        res()
      })
    })
  )
}

export const query = async <T = unknown, V = unknown>(
  sql: string,
  values?: V
): Promise<{ results: ResultSet<T>; fields?: FieldInfo[] }> => {
  return new Promise((res, rej) => {
    connection.query({ sql, values }, (err, results, fields) => {
      if (err) {
        rej(err)
      }
      res({ results, fields })
    })
  })
}

export const init = async () => {
  try {
    await connect()
    await query(TABLE_reservations)
  } catch (err: any) {
    console.error(err)
  }
}
