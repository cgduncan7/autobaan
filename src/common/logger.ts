export enum LogLevel {
  DEBUG,
  INFO,
  ERROR,
}

export class Logger {
  private readonly correlationId: string
  private readonly level: LogLevel

  constructor(correlationId: string, level = LogLevel.ERROR) {
    this.correlationId = correlationId
    this.level = level
  }

  private log(logLevel: LogLevel, message: string, details?: unknown): void {
    if (logLevel < this.level) {
      return
    }

    let levelString
    switch (logLevel) {
      case LogLevel.ERROR:
        levelString = 'ERROR'
        break
      case LogLevel.INFO:
        levelString = 'INFO'
        break
      case LogLevel.DEBUG:
      default:
        levelString = 'DEBUG'
        break
    }

    let fmtString = '[%s] %s: %s'
    const params: Array<unknown> = [this.correlationId, levelString, message]
    if (details) {
      params.push(details)
      fmtString += ' - %O'
    }

    if (logLevel === LogLevel.ERROR) {
      console.error(fmtString, ...params)
    } else {
      console.log(fmtString, ...params)
    }
  }

  public debug(message: string, details?: unknown): void {
    this.log(LogLevel.DEBUG, message, details)
  }

  public info(message: string, details?: unknown): void {
    this.log(LogLevel.INFO, message, details)
  }

  public error(message: string, details?: unknown): void {
    this.log(LogLevel.ERROR, message, details)
  }
}
