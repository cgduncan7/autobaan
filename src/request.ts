import { SQSEvent } from 'aws-lambda'
import { DateTime, Opponent } from './reservation'

export interface IncomingRequest {
  username: string
  password: string
  dateTimes: DateTime[]
  opponent: Opponent
}

export interface ValidationError {
  message: string
  code: number
}

export const validateRequestEvent = (
  event: SQSEvent
): { request?: IncomingRequest; error?: ValidationError } => {
  try {
    const request = validateRequestBody(event.Records[0].body)
    validateRequestDateTimes(request.dateTimes)
    validateRequestOpponent(request.opponent)
    return { request }
  } catch (err: unknown) {
    return { error: { message: 'Invalid request', code: (err as ValidationError).code ?? 0 } }
  }
}

const validateRequestBody = (body?: string): IncomingRequest => {
  if (body === undefined) {
    throw {
      message: 'Invalid request',
      code: 1,
    }
  }

  let jsonBody: IncomingRequest
  try {
    jsonBody = JSON.parse(body)
  } catch (err) {
    throw {
      message: 'Invalid request',
      code: 2,
    }
  }

  const { username, password, dateTimes } = jsonBody
  if (
    !username ||
    username.length < 1 ||
    !password ||
    password.length < 1 ||
    !dateTimes
  ) {
    throw {
      message: 'Invalid request',
      code: 3,
    }
  }

  return jsonBody
}

const validateRequestDateTimes = (dateTimes: DateTime[]): void => {
  const now = new Date()
  for (let i = 0; i < dateTimes.length; i++) {
    const dateTime = dateTimes[i]
    const { year, month, day, timeRange } = dateTime
    const { start, end } = timeRange

    if (
      typeof year !== 'number' ||
      typeof month !== 'number' ||
      typeof day !== 'number' ||
      typeof start.hour !== 'number' ||
      typeof start.minute !== 'number' ||
      typeof end.hour !== 'number' ||
      typeof end.minute !== 'number'
    ) {
      throw {
        message: 'Invalid request',
        code: 4,
      }
    }

    const date = new Date()
    date.setFullYear(year, month - 1, day)
    date.setHours(start.hour, start.minute)

    if (now.getTime() >= date.getTime()) {
      throw {
        message: 'Invalid request',
        code: 5,
      }
    }
  }
}

const validateRequestOpponent = (opponent?: Opponent): void => {
  if (!opponent) return

  const { id, name } = opponent
  if (
    typeof id !== 'string' ||
    typeof name !== 'string' ||
    id.length < 1 ||
    name.length < 1
  ) {
    throw {
      message: 'Invalid request',
      code: 6,
    }
  }
}
