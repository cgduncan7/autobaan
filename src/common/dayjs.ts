import dayjs from 'dayjs'
import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import 'dayjs/locale/nl'
import { asyncLocalStorage } from './logger'

dayjs.extend(isSameOrBefore)
dayjs.extend(utc)
dayjs.extend(timezone)
dayjs.locale('nl')

dayjs.tz.setDefault('Europe/Amsterdam')

const dayjsTz = (
  date?: string | number | Date | dayjs.Dayjs | null | undefined
) => {
  asyncLocalStorage.getStore()?.debug(`<dayjs> timezoning ${date}`)
  return dayjs(date).tz()
}

export default dayjsTz
