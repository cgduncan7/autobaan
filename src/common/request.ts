import dayjs, { Dayjs } from 'dayjs'
import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
dayjs.extend(isSameOrBefore)

import { DateRange, Opponent } from './reservation'

export interface ReservationRequest {
  username: string
  password: string
  dateRanges: {
    start: Dayjs
    end: Dayjs
  }[]
  opponent: Opponent
}

export enum ValidationErrorCode {
  UNDEFINED_REQUEST_BODY = 1,
  INVALID_REQUEST_BODY = 2,
  INVALID_DATE_RANGE = 3,
  INVALID_START_OR_END_DATE = 4,
  INVALID_OPPONENT = 5,
}

export class ValidationError extends Error {
  public readonly code: ValidationErrorCode

  constructor(message: string, code: ValidationErrorCode) {
    super(message)
    this.code = code
  }
}

export const validateRequest = (
  body: string
): ReservationRequest => {
  const request = validateRequestBody(body)
  validateRequestDateRanges(request.dateRanges)
  validateRequestOpponent(request.opponent)
  return request
}

const validateRequestBody = (body?: string): ReservationRequest => {
  if (body === undefined) {
    throw new ValidationError('Invalid request', ValidationErrorCode.UNDEFINED_REQUEST_BODY)
  }

  const jsonBody = transformRequestBody(body)
  const { username, password, opponent, dateRanges } = jsonBody

  if (
    !username ||
    username.length < 1 ||
    !password ||
    password.length < 1 ||
    !dateRanges ||
    dateRanges.length < 1 ||
    (opponent && opponent.id && opponent.id.length < 1) ||
    (opponent && opponent.name && opponent.name.length < 1)
  ) {
    throw new ValidationError('Invalid request', ValidationErrorCode.INVALID_REQUEST_BODY)
  }

  return jsonBody
}

const transformRequestBody = (body: string): ReservationRequest => {
  const json = JSON.parse(body)
  const dateRanges: DateRange[] = json.dateRanges?.map(
    ({ start, end }: { start: string; end: string }): DateRange => {
      return { start: dayjs(start), end: dayjs(end) }
    }
  )
  return {
    username: json.username,
    password: json.password,
    opponent: json.opponent,
    dateRanges,
  }
}

const validateRequestDateRanges = (dateRanges: DateRange[]): void => {
  for (let i = 0; i < dateRanges.length; i++) {
    // checking that both dates are valid
    const { start, end } = dateRanges[i]
    if (!start.isValid() || !end.isValid()) {
      throw new ValidationError('Invalid request', ValidationErrorCode.INVALID_DATE_RANGE)
    }

    // checking that:
    // 1. start occurs after now
    // 2. start occurs before or same as end
    // 3. start and end fall on same YYYY/MM/DD
    if (
      !start.isAfter(dayjs()) ||
      !start.isSameOrBefore(end) ||
      start.format('YYYY MM DD') !== end.format('YYYY MM DD')
    ) {
      throw new ValidationError('Invalid request', ValidationErrorCode.INVALID_START_OR_END_DATE)
    }
  }
}

const validateRequestOpponent = (opponent?: Opponent): void => {
  if (!opponent) return

  const { id, name } = opponent
  if (
    typeof id !== 'string' ||
    typeof name !== 'string' ||
    id.length < 1 ||
    name.length < 1
  ) {
    throw new ValidationError('Invalid request', ValidationErrorCode.INVALID_OPPONENT)
  }
}
