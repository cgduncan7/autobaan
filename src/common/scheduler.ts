import { Dayjs } from 'dayjs'
import { v4 } from 'uuid'

import { Logger, LogLevel } from './logger'
import { Reservation } from './reservation'
import { validateJSONRequest } from './request'
import { reserve } from './reserver'

export interface ScheduledReservation {
  reservation: Reservation
  scheduledFor?: Dayjs
}

export interface SchedulerResult {
  scheduledReservation?: ScheduledReservation
}

export type SchedulerInput = Record<string, unknown>

export const schedule = async (
  payload: SchedulerInput
): Promise<SchedulerResult> => {
  Logger.instantiate('scheduler', v4(), LogLevel.DEBUG)

  Logger.debug('Handling reservation', { payload })
  let reservation: Reservation
  try {
    reservation = await validateJSONRequest(payload)
  } catch (err) {
    Logger.error('Failed to validate request', { err })
    throw err
  }

  Logger.debug('Successfully validated request', {
    reservation: reservation.format(),
  })

  if (!reservation.isAvailableForReservation()) {
    Logger.debug(
      'Reservation date is more than 7 days away; saving for later reservation'
    )

    await Reservation.save(reservation)

    return {
      scheduledReservation: {
        reservation,
        scheduledFor: reservation.getAllowedReservationDate(),
      },
    }
  }

  Logger.info('Reservation request can be performed now')
  await reserve(reservation)
  return {
    scheduledReservation: { reservation },
  }
}
