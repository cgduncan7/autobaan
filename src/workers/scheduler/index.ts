import { Dayjs } from 'dayjs'
import { v4 } from 'uuid'

import { Logger, LogLevel } from '../../common/logger'
import { Reservation } from '../../common/reservation'
import { validateJSONRequest } from '../../common/request'
import { Worker } from '../types'

export interface ScheduledReservation {
  reservation: Reservation
  scheduledFor?: Dayjs
}

export interface SchedulerResult {
  scheduledReservation?: ScheduledReservation
}

export type SchedulerInput = Record<string, unknown>

export const work: Worker<SchedulerInput, SchedulerResult> = async (
  payload: SchedulerInput
): Promise<SchedulerResult> => {
  Logger.instantiate('scheduler', v4(), LogLevel.DEBUG)

  // TODO: obfuscate payload
  Logger.debug('Handling reservation', { payload })
  let reservation: Reservation
  try {
    reservation = await validateJSONRequest(payload)
  } catch (err) {
    Logger.error('Failed to validate request', { err })
    throw err
  }

  Logger.debug('Successfully validated request', {
    reservation: reservation.format(),
  })

  if (!reservation.isAvailableForReservation()) {
    Logger.debug(
      'Reservation date is more than 7 days away; saving for later reservation'
    )
    return {
      scheduledReservation: {
        reservation,
        scheduledFor: reservation.getAllowedReservationDate(),
      },
    }
  }

  Logger.info('Reservation request can be performed now')
  return {
    scheduledReservation: { reservation },
  }
}
