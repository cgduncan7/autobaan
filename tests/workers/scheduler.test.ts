import dayjs from 'dayjs'
import { ValidationError, ValidationErrorCode } from '../../src/common/request'
import { work, SchedulerInput, SchedulerResult } from '../../src/workers/scheduler'

jest.mock('../../src/common/logger')

describe('scheduler', () => {
  test('should handle valid requests within reservation window', async () => {
    const start = dayjs().add(15, 'minutes')
    const end = start.add(15, 'minutes')

    const payload: SchedulerInput = {
      username: "collin",
      password: "password",
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: "123", name: "collin" }
    }

    await expect(work(payload)).resolves
      .toMatchObject<SchedulerResult>({
        scheduledReservationRequest: {
          reservationRequest: {
            username: 'collin',
            password: 'password',
            dateRange: { start, end },
            opponent: { id: '123', name: 'collin' },
          }
        }})
  })

  test('should handle valid requests outside of reservation window', async () => {
    const start = dayjs().add(15, 'days')
    const end = start.add(15, 'minutes')
    const payload: SchedulerInput = {
      username: "collin",
      password: "password",
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: "123", name: "collin" }
    }

    await expect(work(payload)).resolves.toMatchObject<SchedulerResult>({
      scheduledReservationRequest: {
        reservationRequest: {
          username: 'collin',
          password: 'password',
          dateRange: { start, end },
          opponent: { id: '123', name: 'collin' },
        },
        scheduledFor: start.subtract(7, 'days').hour(0).minute(0).second(0).millisecond(0)
      }
    })
  })

  test('should throw error for invalid requests', async () => {
    const start = dayjs().add(15, 'days')
    const end = start.add(15, 'minutes')

    const payload: SchedulerInput = {
      password: "password",
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: "123", name: "collin" }
    }

    await expect(work(payload))
      .rejects
      .toThrowError(new ValidationError('Invalid request', ValidationErrorCode.INVALID_REQUEST_BODY))
  })
})