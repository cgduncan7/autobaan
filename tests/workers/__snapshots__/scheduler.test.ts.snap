// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scheduler should handle valid requests outside of reservation window 1`] = `
Object {
  "scheduledReservation": Object {
    "reservation": Object {
      "booked": false,
      "dateRange": Object {
        "end": Object {
          "$D": 16,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 0,
          "$d": Object {},
          "$m": 15,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
        "start": Object {
          "$D": 16,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 0,
          "$d": Object {},
          "$m": 0,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
      },
      "opponent": Object {
        "id": "123",
        "name": "collin",
      },
      "possibleDates": Array [
        Object {
          "$D": 16,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 0,
          "$d": Object {},
          "$m": 0,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
        Object {
          "$D": 16,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 0,
          "$d": Object {},
          "$m": 15,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
      ],
      "user": Object {
        "password": Any<String>,
        "username": "collin",
      },
    },
    "scheduledFor": Object {
      "$D": 9,
      "$H": 0,
      "$L": "en",
      "$M": 0,
      "$W": 0,
      "$d": Object {},
      "$m": 0,
      "$ms": 0,
      "$s": 0,
      "$x": Object {},
      "$y": 2022,
    },
  },
}
`;

exports[`scheduler should handle valid requests within reservation window 1`] = `
Object {
  "scheduledReservation": Object {
    "reservation": Object {
      "booked": false,
      "dateRange": Object {
        "end": Object {
          "$D": 1,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 6,
          "$d": Object {},
          "$m": 30,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
        "start": Object {
          "$D": 1,
          "$H": 1,
          "$L": "en",
          "$M": 0,
          "$W": 6,
          "$d": Object {},
          "$m": 15,
          "$ms": 0,
          "$s": 0,
          "$x": Object {},
          "$y": 2022,
        },
      },
      "opponent": Object {
        "id": "123",
        "name": "collin",
      },
      "possibleDates": Array [
        "2022-01-01T00:15:00.000Z",
        "2022-01-01T00:30:00.000Z",
      ],
      "user": Object {
        "password": Any<String>,
        "username": "collin",
      },
    },
  },
}
`;
